LanguageExt.Core/DataTypes/Seq/Seq.cs

```diff
@@ -137,20 +138,21 @@ public virtual IEnumerable<A> AsEnumerable()
                 : Seq(this);
 
         /// <summary>
-        /// Match empty sequence, or one item sequence, or multi-item sequence
+        /// Match empty sequence, or one item sequence, or tail of multi-item sequence
         /// </summary>
         /// <typeparam name="B">Return value type</typeparam>
         /// <param name="Empty">Match for an empty list</param>
-        /// <param name="Tail">Match for a non-empty</param>
+        /// <param name="HeadOnly">Match for list with only 1 element</param>
+        /// <param name="Tail">Matches Tail of list with at least 2 elements</param>
         /// <returns>Result of match function invoked</returns>
         public virtual B Match<B>(
             Func<B> Empty,
-            Func<A, B> Head,
+            Func<A, B> HeadOnly,
             Func<Seq<A>, B> Tail) =>
```

Comment:
Leave these as `Head` and `Tail` because it's exactly what's happening.

Reply:
I was just confused because I would intuitively expect 

`Match(Func<B>, Func<A, B>, Func<Seq<A>, B>)`

to give me:
0 elems (= all)
1 elem (= all)
2 or more elems (= all = head included)

this would scale for any number of args and last arg would always be "seq with at least n elems" while args before are "tuple" of 0, 1, ... n-1.

Parameter name Tail is ok (but who looks on arg names to see behaviour, signature should be hint enough) -- IMHO this function could be left out. Having the last arg `Func<A, Seq<A>, B>` can be used instead and has clear meaning. Fits exactly what you wrote in the beginning of your answer.

Reply:
You should look at the implementation, it passes the Tail, not the Head + Tail to the Tail delegate; that's why Tail is the best name here.  As I stated before, the names aren't really important for list matching.  But in this case it should be Tail because it is literally receiving list.Tail as its argument.  

This isn't some obscure use case, it allows the user to work with a complete tail rather than a deconstructed one, which under certain circumstances will be more memory efficient than the variant above that does deconstruct.  That's why I added the variant, to help both potential use-cases.

Reply:
Yes I looked at implementation.

I'm not sure what kind of memory overhead you see. If the function with this signature would be this:

```
        /// <summary>
        /// Match empty sequence, or one item sequence, or tail of multi-item sequence
        /// </summary>
        /// <typeparam name="B">Return value type</typeparam>
        /// <param name="Empty">Match for an empty list</param>
        /// <param name="HeadOnly">Match for list with only 1 element</param>
        /// <param name="More">Matches list with at least 2 elements</param>
        /// <returns>Result of match function invoked</returns>
        public virtual B Match<B>(
            Func<B> Empty,
            Func<A, B> HeadOnly,
            Func<Seq<A>, B> More) =>
            IsEmpty
                ? Empty()
                : this.Tail.IsEmpty
                    ? HeadOnly(this.Head)
                    : More(this);
```

This would match on list size without dropping head or another element in any case as every other Match variant does (yes, this is just one point of view).

If the caller is only interested in Tail he could just call Tail like this:

```
var resultAsString = Seq(1, 2, 3).Match(
    () => "empty", 
    one => $"only: {one}", 
    more=> more.Tail.Apply(xs => "tail: " + string.Join(",", xs)));
```

I don't see why it matters in regard of memory or performance.

Anyway this is all just my humble opinion. My primary goal is to understand design decision to avoid bugs due to misunderstanding and this was something I stumbled on.

You probably don't want to change semantics for this function signature for compatibility reasons, anyway. That's why "drop it" came to my mind. 

I'm happy with current state with cleaned up parameter names.